tests/serve/test_vllm.py

```diff
@@ -0,0 +1,122 @@
+import re
+
+import pytest
+import torch
+
+from outlines.serve.vllm import RegexLogitsProcessor, _patched_apply_logits_processors
+
+
+class MockTokenizer:
+    vocabulary = {
+        **{chr(i): i for i in range(256)},
+        **{"eos": 256},
+    }
+    special_tokens = {"eos"}
+    eos_token_id = 256
+
+    @property
+    def inverse_vocabulary(self):
+        return {v: k for k, v in self.vocabulary.items()}
+
+    def decode(self, token_ids):
+        return "".join([self.inverse_vocabulary[t] for t in token_ids])
+
+    ####
+    # vLLM tokenizer features
+    ####
+    all_special_tokens = list(special_tokens)
+
+    def convert_tokens_to_string(self, token):
+        return token[0]
+
+    def get_vocab(self):
+        return MockTokenizer.vocabulary
+
+
+class MockModel:
+    tokenizer = MockTokenizer()
+
+
+def sample_from_logits(logits):
+    probs = torch.exp(logits) / torch.sum(torch.exp(logits))
+    return torch.multinomial(probs, 1).item()
+
+
+@pytest.mark.parametrize("forget_logits_processor", [True, False])
+def test_time_regexp(forget_logits_processor):
```

Comment:
What is this testing exactly?

Reply:
In vLLM 0.2.6 logits processors would be applied on "random" ray worker. This would result in a regression because a new logits processor had to be created to be created for each worker, causing the loss of previous state.

This test simulates that scenario by creating a new logits processor for every token processed.

However I will have to experiment to determine whether this test and the changes to make it pass are still necessary, as vLLM now applies logits processors on the same worker each time as of vLLM 0.2.7 per https://github.com/outlines-dev/outlines/pull/539#issuecomment-1913752978